### PR TITLE
Laboratory work #1, Merkulova Ekaterina - 19FPL2

### PR DESCRIPTION
**Pull request by katerinamerkulova**

### CODE REVIEW

**Comment by dmitry-uraev** [comment]
Nicely done! Keep on going. Start.py and lint to go

File: lab_1/main.py [hunk 1]
```diff
@@ -4,125 +4,114 @@
...
-    :param tokens: a list of tokens
-    :param stop_words: a list of stop words
-    :return: a list of tokens without stop words
-    e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    stop_words = ['the', 'is']
-    --> ['weather', 'sunny', 'man', 'happy']
-    """
-    pass
+    if isinstance(tokens, list):    # check tokens
+        if isinstance(stop_words, list):    # check stop-words
```
Comment (dmitry-uraev) [inline lab_1/main.py]: consider checking input params without any nesting

Comment (dmitry-uraev) [reply lab_1/main.py]: nesting is bad, especially when it can be avoided

File: lab_1/main.py [hunk 2]
```diff
@@ -4,125 +4,114 @@
...
+            output = [word for word in tokens if word not in stop_words]
+        else:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: you do not need this condition either

File: lab_1/main.py [hunk 3]
```diff
@@ -4,125 +4,114 @@
...
+            output = tokens
+    else:
+        output = []
+    return output
```
Comment (dmitry-uraev) [inline lab_1/main.py]: overall comment for this function (applicable for others) - check input params firstly and then implement main logic

File: lab_1/main.py [hunk 4]
```diff
@@ -4,125 +4,114 @@
...
-    Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
-    :param text: the initial text
-    :return: a list of lowercased tokens without punctuation
-    e.g. text = 'The weather is sunny, the man is happy.'
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
+    try:
+        output = re.sub(r'[^a-z0-9\s]+', '', text.lower()).split()
+    except AttributeError:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: acceptable solution, but be prepared to solve the problem without try-except and re if asked

File: lab_1/main.py [hunk 5]
```diff
@@ -4,125 +4,114 @@
...
-    Calculates frequencies of given tokens
-    :param tokens: a list of tokens without stop words
-    :return: a dictionary with frequencies
-    e.g. tokens = ['weather', 'sunny', 'man', 'happy']
-    --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
-    """
-    pass
+    freqs = {}
+    check = isinstance(tokens, list) and len(tokens) > 0
+    if check and isinstance(tokens[0], str):    # check tokens
```
Comment (dmitry-uraev) [inline lab_1/main.py]: the whole main logic is nested. Consider writing `if not check and not isinstance(tokens[0], str): return {}` and then implement main logic

File: lab_1/main.py [hunk 6]
```diff
@@ -4,125 +4,114 @@
...
+        for word in set(tokens):
+            freqs[word] = tokens.count(word)
+        freqs = dict(sorted(freqs.items(), key=lambda x: x[1], reverse=True))
```
Comment (dmitry-uraev) [inline lab_1/main.py]: acceptable if you can explain lambda function here

File: lab_1/main.py [hunk 7]
```diff
@@ -4,125 +4,114 @@
...
-    right_n = 3
-    --> [['man', 'is'], ['dog, 'cat']]
-    """
-    pass
+    if isinstance(freq_dict, dict) and isinstance(top_n, int):
+        freq_dict = sorted(freq_dict, key=freq_dict.get, reverse=True)
+        output = freq_dict[:top_n]
+    else:
+        output = []
+    return output
```
Comment (dmitry-uraev) [inline lab_1/main.py]: *styling:
1. check input params
2. implement main logic 
NB! No overloading with nesting

File: lab_1/main.py [hunk 8]
```diff
@@ -4,125 +4,114 @@
...
+        rcs = right_context_size
+        check_r = not  isinstance(rcs, bool) and isinstance(rcs, int) and rcs > 0
+        idx = [i for i, x in enumerate(tokens) if x == word]
+        if check_l and check_r:
+            conc = [tokens[i-lcs:i+rcs+1] for i in idx]
+        elif not check_l and check_r:
+            conc = [tokens[i:i+rcs+1] for i in idx]
+        elif check_l and not check_r:
+            conc = [tokens[i-lcs:i+1] for i in idx]
+    return conc
```
Comment (dmitry-uraev) [inline lab_1/main.py]: *styling

File: lab_1/main.py [hunk 9]
```diff
@@ -4,125 +4,114 @@
...
+                    output = sorted(conc, key=lambda x: x[0])
+                elif not left_sort and rcs > 0:
+                    try:
+                        output = sorted(conc, key=lambda x: x[-rcs])
+                    except IndexError:
+                        rcs = len(tokens) - tokens.index(word) - 1
+                        output = sorted(conc, key=lambda x: x[-rcs])
+        except TypeError:
+            output = []
+    return output
```
Comment (dmitry-uraev) [inline lab_1/main.py]: *styling

File: lab_1/main.py [hunk 10]
```diff
@@ -4,125 +4,114 @@
...
+                     word: str,
+                     left_context_size: int,
+                     right_context_size: int,
+                     left_sort: bool) -> list:
+    output = []
+    if isinstance(left_sort, bool):
+        lcs = left_context_size
+        rcs = right_context_size
+        conc = get_concordance(tokens, word, lcs, rcs)
+        try:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: try-except solution is applicable for one-line code where we expect errors, not whole main logic implementation. Consider writing conditions or other checks.

File: lab_1/start.py [hunk 11]
```diff
@@ -1,18 +1,38 @@
...
+    adjacent = main.get_adjacent_words(tokens, 'dog', 2, 0)
+    print('get_adjacent_words', adjacent[:5])
+
+    sorted_concordance = main.sort_concordance(tokens, 'dog', 2, 0, True)
+    print('sort_concordance', sorted_concordance[:5])
+
+    main.write_to_file('', sorted_concordance)
+
+    RESULT = sorted_concordance
+    assert RESULT, 'Concordance not working'
```
Comment (dmitry-uraev) [inline lab_1/start.py:51]: good

**Comment by dmitry-uraev** [comment]
Overall comment: 
Nicely done! Just rewrite some styling issues and be prepared to answer questions on October 3th

**Comment by dmitry-uraev** [comment]
And please, return all docstrings. Do not remove them as they are a part of project

File: lab_1/main.py (hunk 1 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 2 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 3 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 5 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 4 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 6 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 7 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 8 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 9 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/main.py (hunk 10 above)
Comment (katerinamerkulova) [reply lab_1/main.py]: fixed

File: lab_1/start.py (hunk 11 above)
Comment (katerinamerkulova) [reply lab_1/start.py:51]: Thank you a lot!

**Comment by dmitry-uraev** [comment]
done

**Pull request closed without merging**